CARDS-2717: Add belongsToSurvey script for updating old forms

acrowthe opened the pull request:
Adds a new script, addBelongsToSurveyLinks.html.esp, that searches through all existing forms for a given questionnaire and adds belongsToSurvey links from those forms to the relevant survey events form

Testing:
1. Launch in Prems mode
2. Create a bunch of visits, filling out the following questions in the visit information form to generate Survey event forms and patient forms:
  a. Clinic (Make sure to make some inpatient clinics for testing further down)
  b. Date
3. Edit the Clinic on some of those visits to generate additional survey event and patient data forms (eg. update some Inpatient clinics to Inpatient + Integrated Care)
4. Using composum, delete some of the links
5. Navigate to http://localhost:8080/Questionnaires/OAIP.addBelongsToSurveyLinks.html and verify that all the Inpatient forms with deleted links are listed
6. Click the Confirm link at the bottom of the change to apply the changes
7. Using Composum or the CARDS UI, verify that all the Inpatient forms are linked to the appropriate SUrvey Events form
8. Repeat step 7 for all the questionnaires that had deleted links (Replace OAIP in the link with OED, PMOO, YVM, CPESIC, Rehab and/or IC as applicable)

acrowthe committed:
CARDS-2717: Add belongsToSurvey script for updating old forms
Adds a new script, addBelongsToSurveyLinks.html.esp, that searches through all
existing forms for a given questionnaire and adds belongsToSurvey links from those
forms to the relevant survey events form

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/Questionnaire/addBelongsToSurveyLinks.html.esp:
```diff
@@ -0,0 +1,241 @@
+--%>
+<html>
+<head>
+<title>Update last modified date</title>
```
```suggestion
<title>Add missing belongsToSurvey links to legacy forms</title>
```

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/Questionnaire/addBelongsToSurveyLinks.html.esp:
```diff
@@ -0,0 +1,241 @@
+ * @param source The form that shoulf have the belongsToSurvey link added
+ * @param destination The Survey Events form to link to
+ */
+let createLink = function(source, destination) {
```
Why don't you use the LinksUtils API?

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/Questionnaire/addBelongsToSurveyLinks.html.esp:
```diff
@@ -0,0 +1,241 @@
+
+/**
+ * Add a belongsToSurvey link from a form to a survey events form
+ * @param source The form that shoulf have the belongsToSurvey link added
```
```suggestion
 * @param source The form that should have the belongsToSurvey link added
```

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/Questionnaire/addBelongsToSurveyLinks.html.esp:
```diff
@@ -0,0 +1,241 @@
+  let query = "SELECT * FROM [cards:Form] AS f WHERE f.'questionnaire' = '" + currentNode.getIdentifier() + "'";
+  let forms = currentSession.getWorkspace().getQueryManager().createQuery(query, "JCR-SQL2").execute().getNodes();
+  while (forms.hasNext()) {
+    updateForm(forms.next());
```
With the many thousands of forms that are currently in production, this will soon fail because the in-memory session changes will be too big to handle. Add a counter, and every 100 processed forms save the session if it's not in dry-run mode.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/Questionnaire/addBelongsToSurveyLinks.html.esp:
```diff
@@ -0,0 +1,241 @@
+let surveyEventsClinicQuestion = null;
+let belongsToSurveyLinkPath = "/apps/cards/LinkDefinitions/belongsToSurvey";
+let belongsToSurveyLinkType = null;
+
```
```suggestion
let formUtils = sling.getService(Packages.io.uhndata.cards.forms.api.FormUtils);
let linkUtils = sling.getService(Packages.io.uhndata.cards.links.api.LinkUtils);
```

Then you can use them just like in Java.

sdumitriu reviewed (changes requested):
Looks good, just some minor changes needed.

acrowthe committed:
CARDS-2717: Add belongsToSurvey script for updating old forms
- Use link and form utils in update script
- Commit every 100 forms to limit memory use for large updates
- Fix typo and copy paste error

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/libs/cards/Questionnaire/addBelongsToSurveyLinks.html.esp:
```diff
@@ -159,23 +160,16 @@ let updateForm = function(form) {
-    linksContainer = source.addNode("cards:links", 	"cards:Links");
-    linksContainer.setProperty("form", source);
+  pendingChanges++;
+  linkUtils.addLink(source, destination, belongsToSurveyLinkType, "Belongs to Survey");
```
`addLink` saves the session internally, I think there's no need to manually call `session.save` at all.

acrowthe committed:
CARDS-2717: Add belongsToSurvey script for updating old forms
- Clean up session saves as linkUtils handles saving
